### DISCUSSION TITLE
@keydown.shift.tab event followed by @keydown.tab event

### DISCUSSION THREAD

**Discussion by lanius412** [alpinejs/alpine#4173] [7. Bugs (archived)]
I assign ```@keydown.shift.tab``` , ```@keydown.tab``` , ```@keydown.up``` and ```@keydown.down``` to ```button``` 

```@keydown.shift.tab``` and ```@keydown.up``` : focus previous button
```@keydown.tab``` and ```@keydown.down``` : focus next button

When keydown shift+tab,
console output: 
```
prev suggestion
next suggestion
```

Is there something wrong with my code?

```php
@isset($channels)
    <div x-show="showSuggest" x-cloak>
        <ul id="channel-suggestion-list" class="mt-2 py-2 px-3 w-full flex flex-col gap-1 bg-zinc-800 rounded">
            @forelse ($channels as $channel)
                <li>
                    <button type="button"
                        wire:click="$set('search', '{{ $channel['broadcaster_login'] }}')"
                        @click="showSuggest = false"
                        @keydown.shift.tab="prevSuggestion"
                        @keydown.tab="nextSuggestion"
                        @keydown.up="prevSuggestion"
                        @keydown.down="nextSuggestion"
                        class="w-full px-1 text-start rounded hover:bg-zinc-600 focus:outline-none focus:bg-zinc-600" >
                        {{ $channel['broadcaster_login'] }}
                    </button>
                </li>
            @empty
                <li x-cloak>
                    {{ $isEmpty }}
                </li>
            @endforelse
        </ul>
    </div>
@endisset
<script>
    const nextSuggestion = (e) => {
        console.log('next suggestion');
        e.preventDefault();
        const suggestions = document.querySelectorAll('#channel-suggestion-list button');
        const currentSuggestionIndex = [...suggestions].findIndex(elm => e.target === elm);
        if (e.target === document.querySelector('input[name="channel"]') || currentSuggestionIndex+1 === suggestions.length) {
            suggestions[0].focus();
        } else {
            suggestions[currentSuggestionIndex+1].focus();
        }
    }
    const prevSuggestion = (e) => {
        console.log('prev suggestion')
        e.preventDefault();
        const suggestions = document.querySelectorAll('#channel-suggestion-list button');
        const currentSuggestionIndex = [...suggestions].findIndex(elm => e.target === elm);
        if (currentSuggestionIndex === 0) {
            suggestions[suggestions.length-1].focus();
        } else {
            suggestions[currentSuggestionIndex-1].focus();
        }
    }
</script>
```

**Comment by SimoTod** [comment]
The current implementation doesn't prevent the listener to fire if another key is pressed at the same time. Even if you hold shift, you are still pressing tab.

It could be considered a bug by some but changing it could be considered a breaking change by other so you need to pitch it to the maintainer.

**Comment by lanius412** [comment]
Does this mean that it cannot be done like [playground (Vue.js)](https://play.vuejs.org/#eNrNVk1z0zAQ/StCw4ydaSu3A6dgZ2hLoe0BGMoMh4pDaq8dtbJkJDlNJpP/zkrOV2nJEZODI+9bvbd61qy0oKdNw6Yt0CFNbW5E44gF1zYjrkTdaOPIghgoyZKURtckwtSIK65yrawjta1I5vE4ugQpNfmhjSxeRQOu0qSjQyJ8sW4uAYdSkLvWOa2Gpc5bSxZcEXI3zh8qo1tVHOVaajNExuIdV0sSaFZTV+XhqNNuDExRPIYByUYdkQe0BCZ1FUcejwZI8zQOzI1NBW6LOCIFPjJSYEk1KMd+tWDmNyAhd9qcShlHm7p3GB0RqoAZTrxljHmOn6zE0JWPxiBrX9dajmRZRjDWTRcliVeTMXw86Mr3P09z6x9Mgqrc5OgESb1VcTdziSQW/sgPVM8y0cDtl1Iw80v8q1se/+/dOjgJwI49G+eCD8f7vNo6dfCiU083rIO6kWPnNx4h6eRktFiE3b5cpgm+hahQTevI9KjWBciMU8Q5TTqsleEfR1KM0s4N8v4B5oV+VMxOROlwrXfM71I0EWf7Ufw6vA043eb6LJiNc7eT67/WNnd0miadwihNUO7fCZ/1JXzel/CHvoQv+hL+2Jfwp76EL/sSvupL+Pq5cJqEFpYmO52QHlJnsZmXomL3Viu8NITmymmu60ZIMF8aJ7DZczpct11Ox3gxeLwOMWdaOFzH8wnkDy/E7+3Mxzj9asCCmQKnG6w7Hjr44uYzrmIHxD7cSszeA34DPNpaX2OXdoaXDix7Jy9UexWuPkJV3+3FzIGy60X5QsOREfI5xevQ+Z6lb8t9w96ujxq6/A0TMjrX) ?

**Reply by ekwoka** [reply]
Currently the `exact` modifier does not exist.

But it could!

**Reply by lanius412** [reply]
For now, I'll replace it with a Vanilla JS onkeydown event
```
onkeydown="keydown(event)"
{{-- @keydown.shift.tab="prevSuggestion" --}}
{{-- @keydown.tab="nextSuggestion" --}}
{{-- @keydown.up="prevSuggestion" --}}
{{-- @keydown.down="nextSuggestion" --}}

...

const keydown = (e) => {
  // Shift + Tab
  if (e.keyCode === 9 && e.shiftKey) {
      prevSuggestion(e);
  // Tab
  } else if (e.keyCode === 9 && !e.shiftKey) {
      nextSuggestion(e);
  // Arrow Up
  } else if (e.keyCode === 38) {
      prevSuggestion(e);
  // Arrow Down
  } else if (e.keyCode === 40) {
      nextSuggestion(e)
  }
}
```

**Reply by ekwoka** [reply]
No need for it to be using the html attribute.

You can just put that handler on keydown in Alpine.

Specifically, only needed for the tab one.

```
@keydown.tab="$event.shiftKey ? prevSuggestion : nextSuggestion"
```

I am a bit confused about how someone is going to keydown a button...

**Reply by SimoTod** [reply]
you can just have `@keydown.tab="$event.shiftKey ? prevSuggestion(e) : nextSuggestion(e)"`

UPDATE: ekwoka beat me to it 😂
